<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>BVAS Sampler &mdash; bvas  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Simulation" href="simulation.html" />
    <link rel="prev" title="BVAS Selector" href="selector.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            bvas
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="selector.html">BVAS Selector</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">BVAS Sampler</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#bvas.BVASSampler"><code class="docutils literal notranslate"><span class="pre">BVASSampler</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="simulation.html">Simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="map.html">MAP Inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="laplace.html">Laplace Inference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">bvas</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">BVAS Sampler</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/sampler.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bvas-sampler">
<h1>BVAS Sampler<a class="headerlink" href="#bvas-sampler" title="Permalink to this heading">¶</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="bvas.BVASSampler">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bvas.</span></span><span class="sig-name descname"><span class="pre">BVASSampler</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Gamma</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">S</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nu_eff_multiplier</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">explore</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xi_target</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bvas/bvas_sampler.html#BVASSampler"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bvas.BVASSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">MCMCSampler</span></code></p>
<p>MCMC Sampler for Bayesian Viral Allele Selection (BVAS).
Combines a Gaussian diffusion-based likelihood with Bayesian
Variable Selection. Most users will not use this class directly
and will instead use <a class="reference internal" href="selector.html#bvas.BVASSelector" title="bvas.BVASSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">BVASSelector</span></code></a>.</p>
<p>Note that computations will be done using the <cite>device</cite> and <cite>dtype</cite> of the provided
<cite>torch.Tensor`s, i.e. `Y</cite> and <cite>Gamma.</cite> If you would like computations to be done
with a GPU make sure that these tensors are on the GPU. We recommend doing all
computations in 64-bit precision, i.e. <cite>Y.dtype == Gamma.dtype == torch.float64</cite>.</p>
<p>The inputs <cite>Y</cite> and <cite>Gamma</cite> are defined in terms of region-specific allele frequencies
<span class="math notranslate nohighlight">\({\mathbf x}_r(t)\)</span> and region-specific effective population sizes <span class="math notranslate nohighlight">\(\nu_r\)</span> as follows.</p>
<div class="math notranslate nohighlight">
\[ \begin{align}\begin{aligned}&amp;{\mathbf y}_r(t) = {\mathbf x}_r(t + 1) - {\mathbf x}_r(t)\\&amp;\bar{\mathbf{Y}}^\nu \equiv \sum_r \nu_r \sum_t {\mathbf y}_r(t)\\&amp;\Lambda_{r,ab}(t) = x_{r,ab}(t) - x_{r,a}(t) x_{r,b}(t)\\&amp;\bar{\mathbf{\Lambda}}^\nu \equiv \sum_r \nu_r \sum_t {\mathbf \Lambda}_r(t)\end{aligned}\end{align} \]</div>
<p>where <span class="math notranslate nohighlight">\(x_{r,ab}(t)\)</span> denote pairwise allele frequencies in region <span class="math notranslate nohighlight">\(r\)</span>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>Y</strong> (<a class="reference external" href="https://pytorch.org/docs/master/tensors.html#torch.Tensor" title="(in PyTorch vmaster (2.1.0a0+gitbe0b12e ))"><em>torch.Tensor</em></a>) – A vector of shape <cite>(A,)</cite> that encodes integrated alelle frequency
increments for each allele and where <cite>A</cite> is the number of alleles.</p></li>
<li><p><strong>Gamma</strong> (<a class="reference external" href="https://pytorch.org/docs/master/tensors.html#torch.Tensor" title="(in PyTorch vmaster (2.1.0a0+gitbe0b12e ))"><em>torch.Tensor</em></a>) – A matrix of shape <cite>(A, A)</cite> that encodes information about
second moments of allele frequencies.</p></li>
<li><p><strong>S</strong> – Controls the expected number of alleles to include in the model a priori. Defaults to 5.0.
To specify allele-level prior inclusion probabilities provide a A-dimensional <cite>torch.Tensor</cite> of
the form <cite>(h_1, …, h_A)</cite>. If a tuple of positive floats <cite>(alpha, beta)</cite> is provided, the a priori
inclusion probability is a latent variable governed by the corresponding Beta prior so that the sparsity
level is inferred from the data. Note that for a given choice of <cite>alpha</cite> and <cite>beta</cite> the expected number
of alleles to include in the model a priori is given by <span class="math notranslate nohighlight">\(\frac{\alpha}{\alpha + \beta} \times A\)</span>.
We caution that this approach may be a poor choice for very noisy genomic surveillance data.
Also note that the mean number of covariates in the posterior can vary significantly from prior
expectations, since the posterior is in effect a compromise between the prior and the observed data.</p></li>
<li><p><strong>tau</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><em>float</em></a>) – Controls the precision of the coefficients in the prior. Defaults to 100.0.</p></li>
<li><p><strong>nu_eff_multiplier</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><em>float</em></a>) – Additional factor by which to multiply the effective population size, i.e. on top
of whatever was done when computing <cite>Y</cite> and <cite>Gamma</cite>. Defaults to 1.0.</p></li>
<li><p><strong>explore</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><em>float</em></a>) – This hyperparameter controls how greedy the MCMC algorithm is. Defaults to 10.0.
For expert users only.</p></li>
<li><p><strong>xi_target</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><em>float</em></a>) – This hyperparameter controls how often <span class="math notranslate nohighlight">\(h\)</span> MCMC updates are made if <span class="math notranslate nohighlight">\(h\)</span>
is a latent variable. Defaults to 0.2. For expert users only.</p></li>
<li><p><strong>genotype_matrix</strong> (<a class="reference external" href="https://pytorch.org/docs/master/tensors.html#torch.Tensor" title="(in PyTorch vmaster (2.1.0a0+gitbe0b12e ))"><em>torch.Tensor</em></a>) – A matrix of shape <cite>(num_variants, A)</cite> that encodes the genotype
of various viral variants. If included the sampler will compute variant-level growth rates
during inference for the varaints encoded by <cite>genotype_matrix</cite>.
Defaults to None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="selector.html" class="btn btn-neutral float-left" title="BVAS Selector" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="simulation.html" class="btn btn-neutral float-right" title="Simulation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Martin Jankowiak.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>