<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>BVAS Selector &mdash; bvas  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=b3ba4146"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=4825356b"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="BVAS Sampler" href="sampler.html" />
    <link rel="prev" title="BVAS Documentation" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            bvas
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">BVAS Selector</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#bvas.BVASSelector"><code class="docutils literal notranslate"><span class="pre">BVASSelector</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#bvas.BVASSelector.run"><code class="docutils literal notranslate"><span class="pre">BVASSelector.run()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="sampler.html">BVAS Sampler</a></li>
<li class="toctree-l1"><a class="reference internal" href="simulation.html">Simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="map.html">MAP Inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="laplace.html">Laplace Inference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">bvas</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">BVAS Selector</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/selector.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bvas-selector">
<h1>BVAS Selector<a class="headerlink" href="#bvas-selector" title="Permalink to this heading">¶</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="bvas.BVASSelector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bvas.</span></span><span class="sig-name descname"><span class="pre">BVASSelector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">Y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Gamma</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">S</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nu_eff_multiplier</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bvas/bvas_selector.html#BVASSelector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bvas.BVASSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Main analysis class for Bayesian Viral Allele Selection (BVAS).
Combines a Gaussian diffusion-based likelihood with Bayesian
Variable Selection. Uses <a class="reference internal" href="sampler.html#bvas.BVASSampler" title="bvas.BVASSampler"><code class="xref py py-class docutils literal notranslate"><span class="pre">BVASSampler</span></code></a> under the hood to do MCMC
inference.</p>
<p>Usage:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">selector</span> <span class="o">=</span> <span class="n">BVASSelector</span><span class="p">(</span><span class="n">Y</span><span class="p">,</span> <span class="n">Gamma</span><span class="p">,</span> <span class="n">mutations</span><span class="p">,</span> <span class="n">S</span><span class="o">=</span><span class="mf">10.0</span><span class="p">,</span> <span class="n">tau</span><span class="o">=</span><span class="mf">100.0</span><span class="p">)</span>
<span class="n">selector</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">T</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span> <span class="n">T_burnin</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">selector</span><span class="o">.</span><span class="n">summary</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">selector</span><span class="o">.</span><span class="n">growth_rates</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">selector</span><span class="o">.</span><span class="n">stats</span><span class="p">)</span>
</pre></div>
</div>
<p>Note that computations will be done using the <cite>device</cite> and <cite>dtype</cite> of the provided
<cite>torch.Tensor`s, i.e. `Y</cite> and <cite>Gamma.</cite> If you would like computations to be done
with a GPU make sure that these tensors are on the GPU. We recommend doing all
computations in 64-bit precision, i.e. <cite>Y.dtype == Gamma.dtype == torch.float64</cite>.</p>
<p>GPU usage:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">selector</span> <span class="o">=</span> <span class="n">BVASSelector</span><span class="p">(</span><span class="n">Y</span><span class="o">.</span><span class="n">cuda</span><span class="p">()</span><span class="o">.</span><span class="n">double</span><span class="p">(),</span> <span class="n">Gamma</span><span class="o">.</span><span class="n">cuda</span><span class="p">()</span><span class="o">.</span><span class="n">double</span><span class="p">(),</span> <span class="n">mutations</span><span class="p">,</span> <span class="n">S</span><span class="o">=</span><span class="mf">10.0</span><span class="p">,</span> <span class="n">tau</span><span class="o">=</span><span class="mf">100.0</span><span class="p">)</span>
<span class="n">selector</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">T</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span> <span class="n">T_burnin</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
</pre></div>
</div>
<p>The inputs <cite>Y</cite> and <cite>Gamma</cite> are defined in terms of region-specific allele frequencies
<span class="math notranslate nohighlight">\({\mathbf x}_r(t)\)</span> and region-specific effective population sizes <span class="math notranslate nohighlight">\(\nu_r\)</span> as follows.</p>
<div class="math notranslate nohighlight">
\[ \begin{align}\begin{aligned}&amp;{\mathbf y}_r(t) = {\mathbf x}_r(t + 1) - {\mathbf x}_r(t)\\&amp;\bar{\mathbf{Y}}^\nu \equiv \sum_r \nu_r \sum_t {\mathbf y}_r(t)\\&amp;\Lambda_{r,ab}(t) = x_{r,ab}(t) - x_{r,a}(t) x_{r,b}(t)\\&amp;\bar{\mathbf{\Lambda}}^\nu \equiv \sum_r \nu_r \sum_t {\mathbf \Lambda}_r(t)\end{aligned}\end{align} \]</div>
<p>where <span class="math notranslate nohighlight">\(x_{r,ab}(t)\)</span> denote pairwise allele frequencies in region <span class="math notranslate nohighlight">\(r\)</span>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>Y</strong> (<a class="reference external" href="https://pytorch.org/docs/master/tensors.html#torch.Tensor" title="(in PyTorch vmaster (2.1.0a0+gitbe0b12e ))"><em>torch.Tensor</em></a>) – A vector of shape <cite>(A,)</cite> that encodes integrated alelle frequency
increments for each allele and where <cite>A</cite> is the number of alleles.</p></li>
<li><p><strong>Gamma</strong> (<a class="reference external" href="https://pytorch.org/docs/master/tensors.html#torch.Tensor" title="(in PyTorch vmaster (2.1.0a0+gitbe0b12e ))"><em>torch.Tensor</em></a>) – A matrix of shape <cite>(A, A)</cite> that encodes information about
second moments of allele frequencies.</p></li>
<li><p><strong>mutations</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.12)"><em>list</em></a>) – A list of strings of length <cite>A</cite> that encodes the names of the <cite>A</cite> alleles in <cite>Y</cite>.</p></li>
<li><p><strong>S</strong> – Controls the expected number of alleles to include in the model a priori. Defaults to 5.0.
To specify allele-level prior inclusion probabilities provide a A-dimensional <cite>torch.Tensor</cite> of
the form <cite>(h_1, …, h_A)</cite>. If a tuple of positive floats <cite>(alpha, beta)</cite> is provided, the a priori
inclusion probability is a latent variable governed by the corresponding Beta prior so that the sparsity
level is inferred from the data. Note that for a given choice of <cite>alpha</cite> and <cite>beta</cite> the expected number
of alleles to include in the model a priori is given by <span class="math notranslate nohighlight">\(\frac{\alpha}{\alpha + \beta} \times A\)</span>.
We caution that this approach may be a poor choice for very noisy genomic surveillance data.
Also note that the mean number of covariates in the posterior can vary significantly from prior
expectations, since the posterior is in effect a compromise between the prior and the observed data.</p></li>
<li><p><strong>tau</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><em>float</em></a>) – Controls the precision of the coefficients in the prior. Defaults to 100.0.</p></li>
<li><p><strong>nu_eff_multiplier</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><em>float</em></a>) – Additional factor by which to multiply the effective population size, i.e. on
top of whatever was done when computing <cite>Y</cite> and <cite>Gamma</cite>. Defaults to 1.0.</p></li>
<li><p><strong>genotype_matrix</strong> (<a class="reference external" href="https://pytorch.org/docs/master/tensors.html#torch.Tensor" title="(in PyTorch vmaster (2.1.0a0+gitbe0b12e ))"><em>torch.Tensor</em></a>) – A matrix of shape <cite>(num_variants, A)</cite> that encodes the genotype
of various viral variants. If included the sampler will compute variant-level growth rates
during inference for the varaints in <cite>genotype_matrix</cite>.
Defaults to None. If not None, user must also provide <cite>variant_names</cite>.</p></li>
<li><p><strong>variant_names</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.12)"><em>list</em></a>) – A list of names of the variants specified by <cite>genotype_matrix</cite>. Must have the
same length as the leading dimension of <cite>genotype_matrix</cite>. Defaults to None.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="bvas.BVASSelector.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">T</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">T_burnin</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bvas/bvas_selector.html#BVASSelector.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bvas.BVASSelector.run" title="Permalink to this definition">¶</a></dt>
<dd><p>Run MCMC inference for <span class="math notranslate nohighlight">\(T + T_{\rm burn-in}\)</span> iterations.
The leading <span class="math notranslate nohighlight">\(T_{\rm burn-in}\)</span> iterations are discarded. After completion the results
of the MCMC run can be accessed in the <cite>summary</cite>, <cite>growth_rates</cite>, and <cite>stats</cite> attributes.</p>
<p>The <cite>summary</cite> <a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.1.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a> contains six columns.
The first column lists the Posterior Inclusion Probability (PIP) for each covariate.
The second column lists the posterior mean of the coefficient
that corresponds to each covariate. The third column lists the posterior standard deviation for
each coefficient. The fourth and fifth columns are analogous to the second and third columns,
respectively, with the difference that the fourth and fifth columns report conditional posterior
statistics. For example, the fourth column reports the posterior mean of each coefficient
conditioned on the corresponding covariate being included in the model. The sixth column
is the PIP rank.</p>
<p>The <cite>growth_rates</cite> <a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.1.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a> reports estimated relative growth rates for
each variant in <cite>genotype_matrix</cite> if the latter was provided. Note that growth rates are
relative to the wild-type variant with all-zeros genotype.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>T</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a>) – Positive integer that controls the number of MCMC samples that are
generated (i.e. after burn-in/adaptation).</p></li>
<li><p><strong>T_burnin</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a>) – Positive integer that controls the number of MCMC samples that are
generated during burn-in/adaptation.</p></li>
<li><p><strong>seed</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a>) – Random number seed for reproducibility. Defaults to None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="BVAS Documentation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="sampler.html" class="btn btn-neutral float-right" title="BVAS Sampler" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Martin Jankowiak.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>