<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bvas.bvas_selector &mdash; bvas  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js?v=b3ba4146"></script>
        <script src="../../_static/doctools.js?v=888ff710"></script>
        <script src="../../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            bvas
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../selector.html">BVAS Selector</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../sampler.html">BVAS Sampler</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../simulation.html">Simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../map.html">MAP Inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../laplace.html">Laplace Inference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">bvas</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">bvas.bvas_selector</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for bvas.bvas_selector</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">tqdm.contrib</span> <span class="kn">import</span> <span class="n">tenumerate</span>

<span class="kn">from</span> <span class="nn">bvas</span> <span class="kn">import</span> <span class="n">BVASSampler</span>
<span class="kn">from</span> <span class="nn">bvas.containers</span> <span class="kn">import</span> <span class="n">StreamingSampleContainer</span>
<span class="kn">from</span> <span class="nn">bvas.util</span> <span class="kn">import</span> <span class="n">namespace_to_numpy</span>


<span class="k">def</span> <span class="nf">populate_alpha_beta_stats</span><span class="p">(</span><span class="n">container</span><span class="p">,</span> <span class="n">stats</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;h_alpha&#39;</span><span class="p">,</span> <span class="s1">&#39;h_beta&#39;</span><span class="p">,</span> <span class="s1">&#39;h&#39;</span><span class="p">]:</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">container</span><span class="p">,</span> <span class="n">s</span><span class="p">):</span>
            <span class="n">stats</span><span class="p">[</span><span class="s1">&#39;Mean &#39;</span> <span class="o">+</span> <span class="n">s</span><span class="p">]</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">container</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">populate_weight_stats</span><span class="p">(</span><span class="n">selector</span><span class="p">,</span> <span class="n">stats</span><span class="p">,</span> <span class="n">weights</span><span class="p">,</span> <span class="n">quantiles</span><span class="o">=</span><span class="p">[</span><span class="mf">5.0</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">,</span> <span class="mf">20.0</span><span class="p">,</span> <span class="mf">50.0</span><span class="p">,</span> <span class="mf">90.0</span><span class="p">,</span> <span class="mf">95.0</span><span class="p">]):</span>
    <span class="n">q5</span><span class="p">,</span> <span class="n">q10</span><span class="p">,</span> <span class="n">q20</span><span class="p">,</span> <span class="n">q50</span><span class="p">,</span> <span class="n">q90</span><span class="p">,</span> <span class="n">q95</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">weights</span><span class="p">,</span> <span class="n">quantiles</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">s</span> <span class="o">=</span> <span class="s2">&quot;5/10/20/50/90/95:  </span><span class="si">{:.2e}</span><span class="s2">  </span><span class="si">{:.2e}</span><span class="s2">  </span><span class="si">{:.2e}</span><span class="s2">  </span><span class="si">{:.2e}</span><span class="s2">  </span><span class="si">{:.2e}</span><span class="s2">  </span><span class="si">{:.2e}</span><span class="s2">&quot;</span>
    <span class="n">stats</span><span class="p">[</span><span class="s1">&#39;Weight quantiles&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">q5</span><span class="p">,</span> <span class="n">q10</span><span class="p">,</span> <span class="n">q20</span><span class="p">,</span> <span class="n">q50</span><span class="p">,</span> <span class="n">q90</span><span class="p">,</span> <span class="n">q95</span><span class="p">)</span>
    <span class="n">s</span> <span class="o">=</span> <span class="s2">&quot;mean/std/min/max:  </span><span class="si">{:.2e}</span><span class="s2">  </span><span class="si">{:.2e}</span><span class="s2">  </span><span class="si">{:.2e}</span><span class="s2">  </span><span class="si">{:.2e}</span><span class="s2">&quot;</span>
    <span class="n">stats</span><span class="p">[</span><span class="s1">&#39;Weight moments&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">weights</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="o">.</span><span class="n">item</span><span class="p">(),</span> <span class="n">weights</span><span class="o">.</span><span class="n">std</span><span class="p">()</span><span class="o">.</span><span class="n">item</span><span class="p">(),</span>
                                       <span class="n">weights</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="o">.</span><span class="n">item</span><span class="p">(),</span> <span class="n">weights</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">.</span><span class="n">item</span><span class="p">())</span>


<div class="viewcode-block" id="BVASSelector"><a class="viewcode-back" href="../../selector.html#bvas.BVASSelector">[docs]</a><span class="k">class</span> <span class="nc">BVASSelector</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main analysis class for Bayesian Viral Allele Selection (BVAS).</span>
<span class="sd">    Combines a Gaussian diffusion-based likelihood with Bayesian</span>
<span class="sd">    Variable Selection. Uses :class:`BVASSampler` under the hood to do MCMC</span>
<span class="sd">    inference.</span>

<span class="sd">    Usage::</span>

<span class="sd">        selector = BVASSelector(Y, Gamma, mutations, S=10.0, tau=100.0)</span>
<span class="sd">        selector.run(T=2000, T_burnin=1000)</span>
<span class="sd">        print(selector.summary)</span>
<span class="sd">        print(selector.growth_rates)</span>
<span class="sd">        print(selector.stats)</span>

<span class="sd">    Note that computations will be done using the `device` and `dtype` of the provided</span>
<span class="sd">    `torch.Tensor`s, i.e. `Y` and `Gamma.` If you would like computations to be done</span>
<span class="sd">    with a GPU make sure that these tensors are on the GPU. We recommend doing all</span>
<span class="sd">    computations in 64-bit precision, i.e. `Y.dtype == Gamma.dtype == torch.float64`.</span>

<span class="sd">    GPU usage::</span>

<span class="sd">        selector = BVASSelector(Y.cuda().double(), Gamma.cuda().double(), mutations, S=10.0, tau=100.0)</span>
<span class="sd">        selector.run(T=2000, T_burnin=1000)</span>

<span class="sd">    The inputs `Y` and `Gamma` are defined in terms of region-specific allele frequencies</span>
<span class="sd">    :math:`{\mathbf x}_r(t)` and region-specific effective population sizes :math:`\nu_r` as follows.</span>

<span class="sd">    .. math::</span>

<span class="sd">        &amp;{\mathbf y}_r(t) = {\mathbf x}_r(t + 1) - {\mathbf x}_r(t)</span>

<span class="sd">        &amp;\bar{\mathbf{Y}}^\nu \equiv \sum_r \nu_r \sum_t {\mathbf y}_r(t)</span>

<span class="sd">        &amp;\Lambda_{r,ab}(t) = x_{r,ab}(t) - x_{r,a}(t) x_{r,b}(t)</span>

<span class="sd">        &amp;\bar{\mathbf{\Lambda}}^\nu \equiv \sum_r \nu_r \sum_t {\mathbf \Lambda}_r(t)</span>

<span class="sd">    where :math:`x_{r,ab}(t)` denote pairwise allele frequencies in region :math:`r`.</span>

<span class="sd">    :param torch.Tensor Y: A vector of shape `(A,)` that encodes integrated alelle frequency</span>
<span class="sd">        increments for each allele and where `A` is the number of alleles.</span>
<span class="sd">    :param torch.Tensor Gamma: A matrix of shape `(A, A)` that encodes information about</span>
<span class="sd">        second moments of allele frequencies.</span>
<span class="sd">    :param list mutations: A list of strings of length `A` that encodes the names of the `A` alleles in `Y`.</span>
<span class="sd">    :param S: Controls the expected number of alleles to include in the model a priori. Defaults to 5.0.</span>
<span class="sd">        To specify allele-level prior inclusion probabilities provide a A-dimensional `torch.Tensor` of</span>
<span class="sd">        the form `(h_1, ..., h_A)`. If a tuple of positive floats `(alpha, beta)` is provided, the a priori</span>
<span class="sd">        inclusion probability is a latent variable governed by the corresponding Beta prior so that the sparsity</span>
<span class="sd">        level is inferred from the data. Note that for a given choice of `alpha` and `beta` the expected number</span>
<span class="sd">        of alleles to include in the model a priori is given by :math:`\frac{\alpha}{\alpha + \beta} \times A`.</span>
<span class="sd">        We caution that this approach may be a poor choice for very noisy genomic surveillance data.</span>
<span class="sd">        Also note that the mean number of covariates in the posterior can vary significantly from prior</span>
<span class="sd">        expectations, since the posterior is in effect a compromise between the prior and the observed data.</span>
<span class="sd">    :param float tau: Controls the precision of the coefficients in the prior. Defaults to 100.0.</span>
<span class="sd">    :param float nu_eff_multiplier: Additional factor by which to multiply the effective population size, i.e. on</span>
<span class="sd">        top of whatever was done when computing `Y` and `Gamma`. Defaults to 1.0.</span>
<span class="sd">    :param torch.Tensor genotype_matrix: A matrix of shape `(num_variants, A)` that encodes the genotype</span>
<span class="sd">        of various viral variants. If included the sampler will compute variant-level growth rates</span>
<span class="sd">        during inference for the varaints in `genotype_matrix`.</span>
<span class="sd">        Defaults to None. If not None, user must also provide `variant_names`.</span>
<span class="sd">    :param list variant_names: A list of names of the variants specified by `genotype_matrix`. Must have the</span>
<span class="sd">        same length as the leading dimension of `genotype_matrix`. Defaults to None.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">Y</span><span class="p">,</span> <span class="n">Gamma</span><span class="p">,</span> <span class="n">mutations</span><span class="p">,</span> <span class="n">S</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span>
                 <span class="n">tau</span><span class="o">=</span><span class="mf">100.0</span><span class="p">,</span> <span class="n">nu_eff_multiplier</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
                 <span class="n">genotype_matrix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">variant_names</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

        <span class="k">if</span> <span class="n">Y</span><span class="o">.</span><span class="n">ndim</span> <span class="o">!=</span> <span class="mi">1</span> <span class="ow">or</span> <span class="n">Gamma</span><span class="o">.</span><span class="n">ndim</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Y and Gamma must be 1- and 2-dimensional, respectively.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">Y</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="n">Gamma</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">1</span><span class="p">]</span> <span class="ow">or</span> <span class="n">Y</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="n">Gamma</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Y and Gamma must have shapes (A,) and (A, A), respectively, &quot;</span> <span class="o">+</span>
                             <span class="s2">&quot;where A is the number of alleles.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutations</span><span class="p">)</span> <span class="o">!=</span> <span class="n">Y</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="mi">0</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;mutations must be a list of strings with length equal to Y.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">genotype_matrix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="p">(</span><span class="n">genotype_matrix</span><span class="o">.</span><span class="n">ndim</span> <span class="o">!=</span> <span class="mi">2</span> <span class="ow">or</span> <span class="n">genotype_matrix</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">!=</span> <span class="n">Y</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;If genotype_matrix is provided it must have shape (num_variants, A), where A&quot;</span> <span class="o">+</span>
                             <span class="s2">&quot; is the number of alleles.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">genotype_matrix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">variant_names</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span> \
                <span class="p">(</span><span class="n">genotype_matrix</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">variant_names</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;genotype_matrix and variant_names must both be provided or both left unprovided.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">variant_names</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">variant_names</span><span class="p">)</span> <span class="o">!=</span> <span class="n">genotype_matrix</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="mi">0</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;variant names must be a list of strings of length V, &quot;</span> <span class="o">+</span>
                             <span class="s2">&quot;where (V, A) is the shape of genotype_matrix&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">container</span> <span class="o">=</span> <span class="n">StreamingSampleContainer</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutations</span> <span class="o">=</span> <span class="n">mutations</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genotype_matrix</span> <span class="o">=</span> <span class="n">genotype_matrix</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">variant_names</span> <span class="o">=</span> <span class="n">variant_names</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">sampler</span> <span class="o">=</span> <span class="n">BVASSampler</span><span class="p">(</span><span class="n">Y</span><span class="p">,</span> <span class="n">Gamma</span><span class="p">,</span>
                                   <span class="n">nu_eff_multiplier</span><span class="o">=</span><span class="n">nu_eff_multiplier</span><span class="p">,</span> <span class="n">S</span><span class="o">=</span><span class="n">S</span><span class="p">,</span> <span class="n">tau</span><span class="o">=</span><span class="n">tau</span><span class="p">,</span>
                                   <span class="n">explore</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                                   <span class="n">genotype_matrix</span><span class="o">=</span><span class="n">genotype_matrix</span><span class="p">)</span>

<div class="viewcode-block" id="BVASSelector.run"><a class="viewcode-back" href="../../selector.html#bvas.BVASSelector.run">[docs]</a>    <span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">T</span><span class="p">,</span> <span class="n">T_burnin</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Run MCMC inference for :math:`T + T_{\rm burn-in}` iterations.</span>
<span class="sd">        The leading :math:`T_{\rm burn-in}` iterations are discarded. After completion the results</span>
<span class="sd">        of the MCMC run can be accessed in the `summary`, `growth_rates`, and `stats` attributes.</span>

<span class="sd">        The `summary` :class:`pandas.DataFrame` contains six columns.</span>
<span class="sd">        The first column lists the Posterior Inclusion Probability (PIP) for each covariate.</span>
<span class="sd">        The second column lists the posterior mean of the coefficient</span>
<span class="sd">        that corresponds to each covariate. The third column lists the posterior standard deviation for</span>
<span class="sd">        each coefficient. The fourth and fifth columns are analogous to the second and third columns,</span>
<span class="sd">        respectively, with the difference that the fourth and fifth columns report conditional posterior</span>
<span class="sd">        statistics. For example, the fourth column reports the posterior mean of each coefficient</span>
<span class="sd">        conditioned on the corresponding covariate being included in the model. The sixth column</span>
<span class="sd">        is the PIP rank.</span>

<span class="sd">        The `growth_rates` :class:`pandas.DataFrame` reports estimated relative growth rates for</span>
<span class="sd">        each variant in `genotype_matrix` if the latter was provided. Note that growth rates are</span>
<span class="sd">        relative to the wild-type variant with all-zeros genotype.</span>

<span class="sd">        :param int T: Positive integer that controls the number of MCMC samples that are</span>
<span class="sd">            generated (i.e. after burn-in/adaptation).</span>
<span class="sd">        :param int T_burnin: Positive integer that controls the number of MCMC samples that are</span>
<span class="sd">            generated during burn-in/adaptation.</span>
<span class="sd">        :param int seed: Random number seed for reproducibility. Defaults to None.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">enumerate_samples</span> <span class="o">=</span> <span class="n">tenumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sampler</span><span class="o">.</span><span class="n">mcmc_chain</span><span class="p">(</span><span class="n">T</span><span class="o">=</span><span class="n">T</span><span class="p">,</span> <span class="n">T_burnin</span><span class="o">=</span><span class="n">T_burnin</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">),</span>
                                       <span class="n">total</span><span class="o">=</span><span class="n">T</span> <span class="o">+</span> <span class="n">T_burnin</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">t</span><span class="p">,</span> <span class="p">(</span><span class="n">burned</span><span class="p">,</span> <span class="n">sample</span><span class="p">)</span> <span class="ow">in</span> <span class="n">enumerate_samples</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">burned</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">container</span><span class="p">(</span><span class="n">namespace_to_numpy</span><span class="p">(</span><span class="n">sample</span><span class="p">))</span>

        <span class="n">pip</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">container</span><span class="o">.</span><span class="n">pip</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;PIP&quot;</span><span class="p">)</span>
        <span class="n">beta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">container</span><span class="o">.</span><span class="n">beta</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">)</span>
        <span class="n">beta_std</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">container</span><span class="o">.</span><span class="n">beta_std</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;BetaStd&quot;</span><span class="p">)</span>
        <span class="n">conditional_beta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">container</span><span class="o">.</span><span class="n">conditional_beta</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;ConditionalBeta&quot;</span><span class="p">)</span>
        <span class="n">conditional_beta_std</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">container</span><span class="o">.</span><span class="n">conditional_beta_std</span><span class="p">,</span>
                                         <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">mutations</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;ConditionalBetaStd&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">pip</span><span class="p">,</span> <span class="n">beta</span><span class="p">,</span> <span class="n">beta_std</span><span class="p">,</span> <span class="n">conditional_beta</span><span class="p">,</span> <span class="n">conditional_beta_std</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">summary</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;PIP&#39;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary</span><span class="p">[</span><span class="s1">&#39;Rank&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">summary</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">container</span><span class="p">,</span> <span class="s1">&#39;growth_rate&#39;</span><span class="p">):</span>
            <span class="n">growth_rate</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">container</span><span class="o">.</span><span class="n">growth_rate</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;GrowthRate&quot;</span><span class="p">)</span>
            <span class="n">growth_rate_std</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">container</span><span class="o">.</span><span class="n">growth_rate_std</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;GrowthRateStd&quot;</span><span class="p">)</span>
            <span class="n">pango</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">variant_names</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;Variant Name&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">growth_rates</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">growth_rate</span><span class="p">,</span> <span class="n">growth_rate_std</span><span class="p">,</span> <span class="n">pango</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">stats</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">weights</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">container</span><span class="o">.</span><span class="n">_weights</span><span class="p">)</span>
        <span class="n">populate_alpha_beta_stats</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">container</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">stats</span><span class="p">)</span>
        <span class="n">populate_weight_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">stats</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">weights</span><span class="p">)</span></div></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Martin Jankowiak.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>